kubernetes/schema_resources_migrate.go

```diff
@@ -0,0 +1,149 @@
+package kubernetes
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// NOTE these functions are for patching the schemas for resources with containers
+// to use the old schema for the resources block so we can migrate it to the new format
+// without having to duplicate the entire schema
+
+func patchJobTemplatePodSpecWithResourcesFieldV0(m map[string]*schema.Schema) map[string]*schema.Schema {
+	spec := m["spec"].Elem.(*schema.Resource)
+	jobTemplate := spec.Schema["job_template"].Elem.(*schema.Resource)
+	jobSpec := jobTemplate.Schema["spec"].Elem.(*schema.Resource)
+	template := jobSpec.Schema["template"].Elem.(*schema.Resource)
+	template.Schema = patchPodSpecWithResourcesFieldV0(template.Schema)
+	return m
+}
+
+func patchTemplatePodSpecWithResourcesFieldV0(m map[string]*schema.Schema) map[string]*schema.Schema {
+	spec := m["spec"].Elem.(*schema.Resource)
+	template := spec.Schema["template"].Elem.(*schema.Resource)
+	template.Schema = patchPodSpecWithResourcesFieldV0(template.Schema)
+	return m
+}
+
+func patchPodSpecWithResourcesFieldV0(m map[string]*schema.Schema) map[string]*schema.Schema {
+	spec := m["spec"].Elem.(*schema.Resource)
+	initContainer := spec.Schema["init_container"].Elem.(*schema.Resource)
+	initContainer.Schema["resources"].Elem = &schema.Resource{
+		Schema: resourcesFieldV0(),
+	}
+	container := spec.Schema["container"].Elem.(*schema.Resource)
+	container.Schema["resources"].Elem = &schema.Resource{
+		Schema: resourcesFieldV0(),
+	}
+	return m
+}
+
+func upgradeJobTemplatePodSpecWithResourcesFieldV0(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	s, ok := rawState["spec"].([]interface{})
+	if !ok || len(s) == 0 {
+		return rawState, nil
+	}
```

Comment:
This is just a guess on my part, but I think if a user has not specified some of these fields in their config, then they might not be StateUpgraded like they're supposed to.

For example, I had one resource where I didn't specify any resource.requests.limits. I didn't have a `resources{}` block at all in my config. But when I looked at my Terraform state, I saw that `resources{}` (an empty block) had been placed in my state file, and therefore also needed the state upgrade. I think. Again, this is just a hunch.

Reply:
This function looks at the state though, not their config – if it's not in `rawState` then there's nothing there to upgrade.